node/src/middleware.ts

```diff
@@ -0,0 +1,59 @@
+// Copyright 2024 LanceDB Developers.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * Middleware for Remote LanceDB Connection or Table
+ */
+export interface HttpMiddleware {
+  /**
+   * A callback that can be used to instrument the behavior of http requests to remote
+   * tables. It can be used to add headers, modify the request, or even short-circuit
+   * the request and return a response without making the request to the remote endpoint.
+   * It can also be used to modify the response from the remote endpoint.
+   *
+   * @param {RemoteResponse} res - Request ot the remote endpoint
+   * @param {onRemoteRequestNext} next - Callback to advance the middleware chain
+   * @param {MiddlewareContext} ctx - Local context for ths invocation of the middleware
+   */
+  onRemoteRequest(
+    req: RemoteRequest,
+    next: (req: RemoteRequest) => Promise<RemoteResponse>,
+  ): Promise<RemoteResponse>
+};
+
+export enum Method {
+  GET,
+  POST
+}
+
+/**
+ * A LanceDB Remote HTTP Request
+ */
+export interface RemoteRequest {
+  uri: string
+  method: Method
+  headers: Record<string, string>
+  params?: Record<string, string | number>
```

Comment:
Why `string | number`?  Are these partially parsed query params?

Reply:
These are parsed query string params (technically they're query string params before they're serialized). 

I hadn't thought much about the type here .. this is `string | number` b/c that's also what we're passing into client.get
https://github.com/lancedb/lancedb/blob/8f7ef26f5f0644e2cd39b8dd4c21617c6bc99a56/node/src/remote/client.ts#L88

If I were to guess why `number` is allowed here, it's probably to make it easier to pass numberic arguments w/out having to stringify them first.
```js
const limit = 10
client.get("/tables", { limit })
// vs.
client.get("/tables", { limit: `${limit}` })
```
I don't think this is so bad, but I can disallow it being a number if you feel strongly

Reply:
I don't feel strongly.  I guess, in most cases, middleware authors are going to be adding to `req.params` and not reading it.  Should we add `boolean` then so we can handle the full set of JSON primitives?